demo-ling/stock_param_functions.r

```diff
@@ -0,0 +1,47 @@
+## ADDITIONAL FUNCTIONS TO HELP WITH THE SETUP SCRIPTS
+
+## Function to insert stock/species name into g3 parameter
+g3_stock_param <- function(sname, param, lower=NULL, upper=NULL, xby=NULL, sep0="."){
+  
+  nullval <- is.null(lower) + is.null(upper)
+  
+  ## No bounds 
+  if (nullval > 0){
+    if (nullval == 1) warning("One bound is NULL & the other is not NULL, assuming unbounded")
+    tmp <- gadget3:::f_substitute(~g3_param(x), list(x=paste(sname, param, sep=sep0)))
+  }
+  ## Bounds
+  else{
+    lb <- as.numeric(min(lower))
+    ub <- as.numeric(max(upper))
+    tmp <- gadget3:::f_substitute(~bounded(g3_param(x), lb, ub), 
+                                  list(x=paste(sname, param, sep=sep0), lb=lb, ub=ub))
+  }
+  if (!is.null(xby)){
+    tmp <- gadget3:::f_substitute(~y * x, list(y=xby, x=tmp))
+  }
+  return(tmp)
+}
+
+## Function to insert stock/species name into g3 param table (just by age)
+g3_stock_param_table <- function(sname, param, lowage, highage, lower=NULL, upper=NULL, sep0="."){
```

Comment:
Again, pass in the stock object, then you don't need to pass in ``lowage`` / ``highage``, you can pull them out yourself.

Given the name, this should be specific to ``age``. I'd suggest an interface like:-

```
g3_stock_param_table <- function(stock, param, dims, lower = NULL, upper = NULL)
```
Where ``g3_stock_param_table(ling_imm, 'M', c('age'))`` would do what you want, by building up the parameters for ``expand.grid``, one per entry in dims.

The simplest and future-proof way to do this would be ``names(dims) <- dims ; expand.grid(lapply(dims, function (x) ling_imm$dimnames[[x]]))``, but then the ages will be prefixed by "age". Not sure if this is the end of the world or not.

Reply:
For consistency with the `g3_stock_param` and `bounded_param` functions, I broke down the `g3_stock_param_table` into `g3_stock_table` and `bounded_table` functions in commit 87980ee.

`g3_stock_table` creates a stock-specific reference to a table of parameters, as you suggested above the name and ages are taken from the stock object. The `bounded_table` function will add bounds to it from an input list. 

I like your suggestion `names(dims) <- dims ; expand.grid(lapply(dims, function (x) ling_imm$dimnames[[x]]))` and will test this further down the line : )